crates/walrus-service/src/node/db_checkpoint.rs

```diff
@@ -0,0 +1,814 @@
+        } else {
+            time::Instant::now() + Self::CHECKPOINT_CREATION_RETRY_DELAY
+        }
+    }
+
+    /// Calculate the first db_checkpoint time.
+    async fn calculate_first_db_checkpoint_time(
+        db_checkpoint_dir: &Path,
+        db_checkpoint_interval_secs: u64,
+    ) -> Result<time::Instant, DbCheckpointError> {
+        let latest_timestamp = Self::get_latest_db_checkpoint_timestamp(db_checkpoint_dir)?;
+        let now = Utc::now().timestamp();
+        let interval_secs = i64::try_from(db_checkpoint_interval_secs)
+            .map_err(|e| DbCheckpointError::Other(e.into()))?;
+        let next_ts = if let Some(last_ts) = latest_timestamp {
+            std::cmp::max(last_ts + interval_secs, now)
+        } else {
+            now
+        };
+
+        let seconds_from_now = next_ts - now;
+        let duration = std::time::Duration::from_secs(seconds_from_now as u64);
+
+        Ok(time::Instant::now() + duration)
+    }
+
+    /// Get the timestamp of the latest db_checkpoint, if any.
+    pub fn get_latest_db_checkpoint_timestamp(
+        db_checkpoint_dir: &Path,
+    ) -> Result<Option<i64>, DbCheckpointError> {
+        let engine = Self::create_backup_engine(db_checkpoint_dir)?;
+        let backup_info = engine.get_backup_info();
+
+        if backup_info.is_empty() {
+            return Ok(None);
+        }
+
+        let latest_timestamp = backup_info.iter().map(|info| info.timestamp).max();
+
+        Ok(latest_timestamp)
+    }
+
+    /// Create a BackupEngine instance.
+    fn create_backup_engine(db_checkpoint_dir: &Path) -> Result<BackupEngine, DbCheckpointError> {
+        let env = Env::new().map_err(|e| DbCheckpointError::Other(e.into()))?;
+
+        let backup_opts = BackupEngineOptions::new(db_checkpoint_dir)
+            .map_err(|e| DbCheckpointError::Other(e.into()))?;
+
+        BackupEngine::open(&backup_opts, &env).map_err(|e| DbCheckpointError::Other(e.into()))
+    }
+
+    /// Delete old db_checkpoints to maintain the max_db_checkpoints limit.
+    fn purge_old_db_checkpoints(db_checkpoint_dir: &Path, max_db_checkpoints: usize) {
+        if max_db_checkpoints == 0 {
+            return;
+        }
+
+        let Ok(mut engine) = Self::create_backup_engine(db_checkpoint_dir) else {
+            tracing::error!("Failed to create backup engine");
+            return;
+        };
+        let result = engine
+            .purge_old_backups(max_db_checkpoints)
+            .map_err(|e| DbCheckpointError::Other(anyhow::anyhow!("Purge error: {}", e)));
+
+        match result {
+            Ok(_) => tracing::info!(
+                "purged old db_checkpoints, keeping {} most recent",
+                max_db_checkpoints
+            ),
+            Err(e) => tracing::error!(?e, "Failed to purge old db_checkpoints"),
+        }
+    }
+
+    /// Restore from the most recent backup.
+    pub async fn restore_latest(
```

Comment:
Could you add some comments or documentation clarifying that the `restore_latest` function is intended for **offline** use (i.e., the node must be stopped when restoring the database)? This is standard for RocksDB backups but worth making explicit. I only see it currently being used in a test.